rye/src/cli/publish.rs

```diff
@@ -181,25 +123,148 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
     Ok(())
 }
 
-fn prompt_for_token() -> Result<String, Error> {
+#[derive(Debug)]
+struct RepositoryConfig {
+    username: Option<String>,
+    password: Option<Secret<String>>,
+    repository_url: Option<Url>,
+}
+
+/// Resolve repository configuration needed for publishing packages.
+///
+/// 1. Resolve any configuration from ~/.rye/credentials using '--repository' (default is 'pypi')
+/// 2. Override any resolved configuration with CLI.
+/// 3. Prompt user for missing options (skips with '--yes').
+/// 4. Finally, if no token is found, we can fallback to (username, none, url) for keyring backends.
+///
+/// We try to do our best to preserve a 'pypi' defaults. If --skip-save-credentials isn't provided
+/// Rye will save the resolved credentials for '--repository' in ~/.rye/credentials.
+fn resolve_repository_config(cmd: &Args) -> Result<RepositoryConfig, Error> {
+    let mut config = RepositoryConfig {
+        username: None,
+        password: None,
+        repository_url: None,
+    };
+
+    let mut rye_credentials = get_credentials()?;
+    let repo_credentials = rye_credentials
+        .entry(&cmd.repository)
+        .or_insert(Item::Table(Table::new()));
+
+    // Attempt to resolve url from credentials if we aren't provided with one.
+    if let Some(url) = repo_credentials
+        .get("repository-url")
+        .map(Item::to_string)
+        .map(escape_string)
+    {
+        config.repository_url = Some(Url::parse(&url)?);
+    }
+
+    if cmd.repository_url.is_some() {
+        config.repository_url = cmd.repository_url.clone();
```

Comment:
I felt that adding these few conditional clones would be fine to make things easier to follow. Let me know if you're unhappy with it. I'm always looking to learn to better borrow management, especially if this isn't necessary.